### PR TITLE
chore(core): refactor metadata infrastructure

### PR DESCRIPTION
**Pull request by bluestreak01**
Accessing metadata was confusing and mistake prone endeavor. This PR attempts to split metadata into:

- sequencer metadata (own pool)
- table metadata (own pool)
- legacy metadata (uses either of the pools depending on WAL or non-WAL statu of the table)

Ent PR: https://github.com/questdb/questdb-enterprise/pull/296

### CODE REVIEW

File: core/src/main/java/io/questdb/cairo/CairoEngine.java [hunk 1]
```diff
@@ -429,6 +429,25 @@ public MessageBus getMessageBus() {
...
+    }
+
+        /**
+         * Retrieves up-to-date table metadata regardless of table type.
+         *
+         * @param tableToken table token
+         * @param desiredVersion version of table metadata used previously if consistent metadata reads are required
+         * @return returns {@link io.questdb.cairo.wal.seq.SequencerMetadata} for WAL tables and {@link TableMetadata}
+         * for non-WAL, which would be metadata of the {@link TableReader}
+         */
```
Comment (ideoma) [inline core/src/main/java/io/questdb/cairo/CairoEngine.java:443]: nit: formatting indentation is off

File: core/src/main/java/io/questdb/cairo/CairoEngine.java [hunk 2]
```diff
@@ -498,33 +517,43 @@ public TableReader getReaderWithRepair(TableToken tableToken) {
...
-                throw ex;
-            }
-            return metadata;
+            return validateDesiredMetadataVersion(
+                    tableToken,
+                    sequencerMetadataPool.get(tableToken),
+                    desiredVersion
+            );
         } catch (CairoException e) {
             tryRepairTable(tableToken, e);
```
Comment (ideoma) [inline core/src/main/java/io/questdb/cairo/CairoEngine.java:549]: No point in trying to repair the table, WAL table sequencer metadata will not be repaired by opening TableWriter.

File: core/src/main/java/io/questdb/cairo/CairoEngine.java [hunk 3]
```diff
@@ -1262,6 +1315,21 @@ private void validNameOrThrow(CharSequence tableName) {
         }
     }
 
+    private TableMetadata validateDesiredMetadataVersion(TableToken tableToken, TableMetadata metadata, long desiredVersion) {
+        if (desiredVersion != TableUtils.ANY_TABLE_VERSION && metadata.getMetadataVersion() != desiredVersion) {
+            final TableReferenceOutOfDateException ex = TableReferenceOutOfDateException.of(
+                    tableToken,
+                    metadata.getTableId(),
```
Comment (ideoma) [inline core/src/main/java/io/questdb/cairo/CairoEngine.java:1322]:
```suggestion
                    tableToken.getTableId(),
```

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 46 / 56 (82.14%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetadata.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|22|30|73.33%|
|:large_blue_circle:|io/questdb/cairo/pool/TableMetadataPool.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalApplySqlExecutionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlExecutionContext.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/SequencerMetadataPool.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/ShowTablesFunctionFactory.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|2|2|100.00%|

File: core/src/main/java/io/questdb/cairo/CairoEngine.java (hunk 3 above)
Comment (bluestreak01) [reply core/src/main/java/io/questdb/cairo/CairoEngine.java]: fixed in https://github.com/questdb/questdb/pull/4044/commits/1b26f6db43a998a89649ecb8f103a9fc2740ad43

File: core/src/main/java/io/questdb/cairo/CairoEngine.java (hunk 2 above)
Comment (bluestreak01) [reply core/src/main/java/io/questdb/cairo/CairoEngine.java:549]: fixed in https://github.com/questdb/questdb/pull/4044/commits/1b26f6db43a998a89649ecb8f103a9fc2740ad43

File: core/src/main/java/io/questdb/cairo/CairoEngine.java (hunk 1 above)
Comment (bluestreak01) [reply core/src/main/java/io/questdb/cairo/CairoEngine.java]: formatted in https://github.com/questdb/questdb/pull/4044/commits/1b26f6db43a998a89649ecb8f103a9fc2740ad43

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 46 / 54 (85.19%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetadata.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|22|28|78.57%|
|:large_blue_circle:|io/questdb/cairo/pool/TableMetadataPool.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalApplySqlExecutionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlExecutionContext.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/SequencerMetadataPool.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/ShowTablesFunctionFactory.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|2|2|100.00%|

**Review by ideoma — APPROVED** [review]

**Pull request merged**